README.md
# PMDfinder

### A Python Package to identify partially methylation domains (PMDs) from whole-genome bisulfite sequencing (WGBS) data

By Yihui Zhu

## Language
Python 3

## Running PMDfinder:
First prepare WGBS data in BED file format.

## Installation
Make local copy of the package from Github and change directory in Terminal to the local package folder
```
$ git clone https://github.com/Yihui-Zhu/PMDfinder.git
$ cd PMDfinder
$ pip install .
```

## Generic Usage Example
```
$ python
>>> from PMDfinder import findPMD
>>> findPMD("tests/DSS_chr22_files", "tests/cpgIslandExt.hg38.bed", "tests/output_meth.bed", "tests/output_GRanges.bed", 0.9, 20, "tests/meth_plot.png")
```

## Function
`findPMD` is the main function of the PMDfinder package with methylome as input with output as PMD location.
```
findPMD(directory, CpG_Island_path, outputpath1, outputpath2, percentile=0.9, cutoff=20, plotpath='tests/meth_plot.png')
```
- `directory`: input BED files directory path.
- `CpG_Island_path`: input CpG Island BED files path.
- `outputpath1`: the output bed file path.
- `outputpath2`: the output grange file path.
- `percentile`: Percent of samples per CpG coverage, values range from 0 to 1.
- `cutoff`: minimum number of reads per CpG site
- `plotpath`: the output figure path. 

`generateFigure` is the plotting function of PMDfinder package that enable plot results from the output methylation file
```
generateFigure(output_methylation, path)
```
- `output_methylation`: BED file.
- `path`: the output figure path.

## Input BED file format:
Input files should be split by each chromosome. The format should be in BED which is a tab-delimited text file with 4 columns: chromosome (chr), position (pos), total reads (N), and methylated reads (X) (see below).

        chr	        pos	        N	X
        chr22	21971583	39	26
        chr22	21971603	34	31
        chr22	21971653	26	25
        chr22	21971662	24	24
        chr22	21971701	41	40

## Input Directory:
Store all samples BED files inside one directory. For example: `tests/DSS_chr22_files`

## Input CpG island:
For `hg38`, `cpgIslandExt` table was extracted from UCSC and used [BEDOPS](http://bedops.readthedocs.io/en/latest/content/reference/file-management/sorting/sort-bed.html) `sort-bed` for BED file.
Command line adapted from [biostars page](https://www.biostars.org/p/236141/). 
```
wget -qO- http://hgdownload.cse.ucsc.edu/goldenpath/hg38/database/cpgIslandExt.txt.gz \
   | gunzip -c \
   | awk 'BEGIN{ OFS="\t"; }{ print $2, $3, $4, $5$6, substr($0, index($0, $7)); }' \
   | sort-bed - \
   > cpgIslandExt.hg38.bed
```
### Example cpgIslandExt.hg38.bed:
```
chr22	48500563	48500779	CpG:17	216	17	134	15.7	62	0.82
chr22	48560078	48560294	CpG:15	216	15	139	13.9	64.4	0.68
chr22	48575045	48577135	CpG:290	2090	290	1549	27.8	74.1	1.01
chr22	48581364	48581571	CpG:16	207	16	128	15.5	61.8	0.84
chr22	48623105	48623505	CpG:47	400	47	280	23.5	70	1
```

### PMDfinder workflow plot:
<div align="center"><img src="https://github.com/Yihui-Zhu/PMDfinder/blob/main/Figures/PMDfinder_workflow.png" alt="PMDfinder workflow plot" width="640" height="1300"></div>

### Example output:
<div align="center"><img src="https://github.com/Yihui-Zhu/PMDfinder/blob/main/Figures/PMDfinder_output.png" alt="PMDfinder output plot" width="2200" height="500"></div>

Each dot represents the percent methylation level of an individual CpG site. The red line on the top represent the identification of PMD. The red line on the bottom represent the identification of Non-PMD.

### Example output_meth BED:
        chr	        pos	        meth_ratio	PMD_predict
        chr22	49175149	0.5735586481113321	PMD
        chr22	49175157	0.3700707785642063	PMD
        chr22	49175225	0.7020905923344948	PMD
        chr22	49175246	0.6876404494382022	PMD
        chr22	49175265	0.4755784061696658	PMD

### Example GRanges BED:
        chr	        start	        end	        status
        chr22	50169395	50171197	CpG_Island
        chr22	50171248	50173818	PMD
        chr22	50173823	50176803	Non-PMD
        chr22	50176850	50179470	CpG_Island
        chr22	50179513	50184700	Non-PMD

## UC Davis Tristan Server Example
Install TensorFlow and activate it.
```
conda create -n tf-gpu tensorflow-gpu
conda activate tf-gpu
```

## Version
v0.3.0  

## Dataset
- Zhu, Yihui, J. Antonio Gomez, Benjamin I. Laufer, Charles E. Mordaunt, Julia S. Mouat, Daniela C. Soto, Megan Y. Dennis et al. "Placental methylome reveals a 22q13. 33 brain regulatory gene locus associated with autism." Genome biology 23, no. 1 (2022): 1-32. **doi:** https://doi.org/10.1186/s13059-022-02613-1
- Datasets supporting the conclusions are available in the Gene Expression Omnibus repository (GEO): https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE178206


CHANGELOG.txt
Change Log
==========

0.1.0 (06/01/2022)
-------------------
- One sample Release

0.2.0 (07/06/2022)
-------------------
- Take one directory input files
- Added data preprocessing 
- Output figure

0.3.0 (07/12/2022)
-------------------
- Filter on CpG Island


PMDfinder/PMDfinder.py
# Load Packages
import pandas as pd
import numpy as np
# import random
from matplotlib import pyplot as plt
from sklearn.model_selection import LeaveOneOut
from sklearn import preprocessing
from sklearn.metrics import accuracy_score, precision_score, recall_score
from sklearn.model_selection import train_test_split
from sklearn.cluster import KMeans
import tensorflow as tf
from tensorflow.keras import layers, losses
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.models import Model           


class Autoencoder(Model):
    def __init__(self, latent_dim):
        super(Autoencoder, self).__init__()
        self.latent_dim = latent_dim   
        self.encoder = tf.keras.Sequential([
        layers.Dense(latent_dim, activation=layers.LeakyReLU(alpha=0.3)),
        ])
        self.decoder = tf.keras.Sequential([
        layers.Dense(1024, activation=layers.LeakyReLU(alpha=0.3))
        ])

    def call(self, x):
        encoded = self.encoder(x)
        decoded = self.decoder(encoded)
        return decoded  

def findPMD(filepath, outputpath):
    """
        findPMD(filepath, outputpath)
    The main function in PMDfinder.
    
    * filepath: input BED file path.
    * outputpath: the output file path.
    """
    # load DSS methylation data
    methylation = pd.read_csv(filepath)

    # store the location and the percent methylation
    meth_ratio = methylation["M"]/methylation["T"]
    geno_pos = methylation["start"]

    ### Data Conversion
    # convert methylation ratio to PMD/non-PMD level (y=4x(1-x))
    def methRatio2PMDLevel(meth_ratio):
        n = len(meth_ratio)
        PMD_level = [0]*n
        for i in range(n):
            PMD_level[i] = 4 * meth_ratio[i] * (1 - meth_ratio[i])
        return PMD_level

    PMD_level = methRatio2PMDLevel(meth_ratio)

    ### Sequential Data Matrix
    # Extract sequenctial feature by sliding window
    N = len(PMD_level)
    X = np.zeros((N-1023, 1024))
    for i in range(N-1023):
        X[i, :] = PMD_level[i:i+1024]

    ### Autoencoder
    # latent is the last variable
    latent_dim = 8

    m = Autoencoder(latent_dim)
    m.compile(optimizer='adam', loss=losses.MeanSquaredError())

    # fit the model
    m.fit(X, X, epochs=5, shuffle=True)

    # get the encoded PMD
    encoded_slicing_PMD = m.encoder(X).numpy()

    ### k-means
    kmeans = KMeans(n_clusters=2, random_state=22).fit(encoded_slicing_PMD)
    final_result = kmeans.labels_

    ### Post-processing steps
    ## Remove PMD that is less than 51 bp length
    assign1 = [] # index for the location equal to 1
    for i in range(len(final_result)):
        if final_result[i] == 1:
            assign1.append(i)

    break_pts1 = [] # index for the break point, the next equal to 1 is more than 1bp
    for i in range(1, len(assign1)):
        if assign1[i] - assign1[i-1] > 1:
            break_pts1.append(i)    

    # small_PMD_intervals: identify region that is close with each other 
    small_PMD_intervals = []
    for i in range(1, len(break_pts1)):
        if assign1[break_pts1[i]-1] - assign1[break_pts1[i-1]] + 1 < 101:
            small_PMD_intervals.append(i)

    # change the PMD interval with less than 51 to Non-PMD
    for interval in small_PMD_intervals:
        final_result[assign1[break_pts1[interval-1] : break_pts1[interval]-1]] == 0

    ## Merge PMD that is less than 101 bp from the next one
    # This need to check the non-PMD region length
    assign2 = []
    for i in range(len(final_result)):
        if final_result[i] == 0:
            assign2.append(i)

    break_pts2 = []
    for i in range(1, len(assign2)):
        if assign2[i] - assign2[i-1] > 1:
            break_pts2.append(i)

    # small non_PMD intervals
    small_non_PMD_intervals = []
    for i in range(1, len(break_pts2)):
        if assign2[break_pts2[i]-1] - assign2[break_pts2[i-1]] + 1 < 101:
            small_non_PMD_intervals.append(i)

    # change the PMD interval with less than 51 to Non-PMD
    for interval in small_non_PMD_intervals:
        final_result[assign2[break_pts2[interval-1] : break_pts2[interval]-1]] == 0

    np.savetxt(outputpath, final_result, delimiter=',')


PMDfinder/main.py
# Load Packages
import pandas as pd
import numpy as np
import random
import os
import matplotlib
matplotlib.use('TKAgg')
from matplotlib import pyplot as plt
from sklearn.model_selection import LeaveOneOut
from sklearn import preprocessing
from sklearn.metrics import accuracy_score, precision_score, recall_score
from sklearn.model_selection import train_test_split
from sklearn.cluster import KMeans
import tensorflow as tf
tf.enable_eager_execution()
tf.executing_eagerly()
from tensorflow.keras import layers, losses
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.models import Model           
configuration = tf.compat.v1.ConfigProto(device_count={"GPU": 0})
session = tf.compat.v1.Session(config=configuration)
random.seed(12)

class Autoencoder(Model):
    def __init__(self, latent_dim):
        super(Autoencoder, self).__init__()
        self.latent_dim = latent_dim   
        self.encoder = tf.keras.Sequential([
        layers.Dense(latent_dim, activation=layers.LeakyReLU(alpha=0.3)),
        ])
        self.decoder = tf.keras.Sequential([
        layers.Dense(1024, activation=layers.LeakyReLU(alpha=0.3))
        ])

    def call(self, x):
        encoded = self.encoder(x)
        decoded = self.decoder(encoded)
        return decoded  

def findPMD(directory, CpG_Island_path, outputpath1, outputpath2, percentile=0.95, cutoff=20, plotpath='tests/meth_plot.png'):
    """
        findPMD(directory, CpG_Island_path, outputpath1, outputpath2, percentile=0.95, cutoff=20, plotpath='tests/meth_plot.png')
    The main function in PMDfinder.
    
    * directory: input BED files directory path.
    * CpG_Island_path: input CpG Island BED files path.
    * outputpath1: the output bed file path.
    * outputpath2: the output grange file path.
    * percentile: Percent of samples per CpG coverage, values range from 0 to 1.
    * cutoff: minimum number of reads per CpG site
    * plotpath: the output figure path. 
    """
    
    # load DSS methylation data
    meth_ratios = []
    for filename in os.listdir(directory):
        filepath = os.path.join(directory, filename)
        methylation = pd.read_csv(filepath, sep='\t', comment='t', header=0, low_memory=False)
        # store the location and the percent methylation
        a = list(map(float, methylation['X']))
        b = list(map(float, methylation['N']))
        # ratio = [i / j for i, j in zip(a, b)]
        twonum = [(i, j) for i, j in zip(a, b)]
        geno_pos = list(map(int, methylation['pos']))
        meth_ratio_dict = dict(zip(geno_pos, twonum))
        meth_ratios.append(meth_ratio_dict)

    ### Data pre-processing
    def methRatioPreprocessing(meth_ratios, percentile, cutoff):
        union_pos = set()
        for d in meth_ratios:
            union_pos |= set(d.keys())
        
        union_pos = sorted(list(union_pos))
        meth_ratio = {}
        for pos in union_pos:
            temp = []
            for i, ratio in enumerate(meth_ratios):
                if pos in ratio:
                    temp.append(i)
            if len(temp) > percentile * len(meth_ratios):
                count = sum([meth_ratios[j][pos][1] for j in temp])
                if count > cutoff:
                    is_meth = sum([meth_ratios[j][pos][0] for j in temp])
                    meth_ratio[pos] =  is_meth / count

        return meth_ratio

    ### Data Conversion
    # convert methylation ratio to PMD/non-PMD level (y=4x(1-x))
    def methRatio2PMDLevel(meth_ratio):
        n = len(meth_ratio)
        PMD_level = [0]*n
        for i in range(n):
            PMD_level[i] = 4 * meth_ratio[i] * (1 - meth_ratio[i])
        return PMD_level

    position2ratio = methRatioPreprocessing(meth_ratios, percentile, cutoff)
    position = list(position2ratio.keys())
    meth_ratio = list(position2ratio.values())
    PMD_level = methRatio2PMDLevel(meth_ratio)

    ### Sequential Data Matrix
    # Extract sequenctial feature by sliding window
    N = len(PMD_level)
    X = np.zeros((N-1023, 1024))
    for i in range(N-1023):
        X[i, :] = PMD_level[i:i+1024]
    
    X = X.astype(np.float32)

    ### Autoencoder
    # latent is the last variable
    latent_dim = 8

    m = Autoencoder(latent_dim)
    m.compile(optimizer='adam', loss=losses.MeanSquaredError())

    # fit the model
    m.fit(X, X, epochs=3, shuffle=True)

    # get the encoded PMD
    encoded_slicing_PMD = m.encoder(X).numpy()

    ### k-means
    kmeans = KMeans(n_clusters=2, random_state=22).fit(encoded_slicing_PMD)
    final_result = kmeans.labels_

    ### Post-processing steps
    ## Remove PMD that is less than 101 bp length
    assign1 = [] # index for the location equal to 1
    for i in range(len(final_result)):
        if final_result[i] == 1:
            assign1.append(i)

    break_pts1 = [0] # index for the break point, the next equal to 1 is more than 1bp
    for i in range(1, len(assign1)):
        if assign1[i] - assign1[i-1] > 1:
            break_pts1.append(i)    
    
    # small_PMD_intervals: identify region that is close with each other 
    small_PMD_intervals = []
    for i in range(1, len(break_pts1)):
        if assign1[break_pts1[i]-1] - assign1[break_pts1[i-1]] + 1 < 6:
            small_PMD_intervals.append(i)

    # change the PMD interval with less than 6 to Non-PMD
    for interval in small_PMD_intervals:
        final_result[assign1[break_pts1[interval-1] : break_pts1[interval]]] = 0

    ## Merge PMD that is less than 6 bp from the next one
    # This need to check the non-PMD region length
    assign2 = []
    for i in range(len(final_result)):
        if final_result[i] == 0:
            assign2.append(i)

    break_pts2 = [0]
    for i in range(1, len(assign2)):
        if assign2[i] - assign2[i-1] > 1:
            break_pts2.append(i)

    # small non_PMD intervals
    small_non_PMD_intervals = []
    for i in range(1, len(break_pts2)):
        if assign2[break_pts2[i]-1] - assign2[break_pts2[i-1]] + 1 < 6:
            small_non_PMD_intervals.append(i)

    # change the PMD interval with less than 51 to Non-PMD
    for interval in small_non_PMD_intervals:
        final_result[assign2[break_pts2[interval-1] : break_pts2[interval]]] = 1
    
    # file output
    output_methylation = pd.DataFrame(['chr22'] * (len(meth_ratio)-1023), columns=['chr'])
    output_methylation.loc[:, 'pos'] = position[:len(meth_ratio)-1023]
    output_methylation.loc[:, 'meth_ratio'] = meth_ratio[:len(meth_ratio)-1023]

    one_avg = np.mean([PMD_level[x] for x in range(len(final_result)) if final_result[x] == 1])
    zero_avg = np.mean([PMD_level[x] for x in range(len(final_result)) if final_result[x] == 0])
    print("k-means label 0 PMD_level avg:", zero_avg, "; label 1 avg:", one_avg)

    output_methylation.loc[:, 'PMD_predict'] = pd.DataFrame(final_result)[0].map({1: 'PMD', 0: 'Non-PMD'}) if one_avg > zero_avg else pd.DataFrame(final_result)[0].map({1: 'Non-PMD', 0: 'PMD'})
    CpG_Islands = pd.read_csv(CpG_Island_path, sep='\t', header=None, usecols=[0,1,2])
    CpG_Islands = CpG_Islands[CpG_Islands[0] == 'chr22'] 
    CpG_Islands_rows = []
    ind = 0
    for i, pos in enumerate(output_methylation['pos']):
        if pos > CpG_Islands.iloc[ind, 2] and ind < len(CpG_Islands.index):
            ind += 1
        if CpG_Islands.iloc[ind, 1] <= pos <= CpG_Islands.iloc[ind, 2]:
            CpG_Islands_rows.append(i)
    
    output_methylation_reduced = output_methylation.drop(CpG_Islands_rows)
    output_methylation_reduced.to_csv(outputpath1, sep='\t', index=False, header=True)

    # output grange file
    df = pd.DataFrame(columns = ['chr', 'start', 'end', 'status'])

    CpG_Islands_rows = set(CpG_Islands_rows)
    ncols = len(output_methylation)
    i, j = 0, 0
    while i < ncols:
        if j == ncols:
            df = df.append({'chr': output_methylation.iloc[i, 0], 'start': output_methylation.iloc[i, 1], 'end': output_methylation.iloc[j-1, 1], 'status': ti}, ignore_index=True)
            break

        ti = output_methylation.iloc[i, 3] if i not in CpG_Islands_rows else 'CpG_Island'
        tj = output_methylation.iloc[j, 3] if j not in CpG_Islands_rows else 'CpG_Island'
        if tj == ti:
            j += 1
        else:
            df = df.append({'chr': output_methylation.iloc[i, 0], 'start': output_methylation.iloc[i, 1], 'end': output_methylation.iloc[j-1, 1], 'status': ti}, ignore_index=True)
            i = j

    df.to_csv(outputpath2, sep='\t', index = False, header=True)
    # print(df)
    generateFigure(output_methylation_reduced, plotpath)

    print("Finished PMDfinder!")

    # np.savetxt(outputpath1, outputpath2, final_result, delimiter=',')

def generateFigure(df, path):
    """
        generateFigure(output_methylation, path)
    The plotting function in PMDfinder.
    
    * output_methylation: BED file.
    * path: the output figure path.
    """
    fig = plt.figure(figsize=(40, 15), dpi=100)

    plt.plot(df['pos'], df['meth_ratio'], 'o', color='black', label='meth ratio', ms = 1)
    plt.plot(df['pos'], df['PMD_predict'] == 'PMD', 'o', color='red')

    current_values = plt.gca().get_xticks()
    plt.gca().set_xticklabels(['{:.0f}'.format(x) for x in current_values])

    fig.savefig(path)